src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -59,19 +59,170 @@ int32_t GlobalizationNative_CompareStringNative(const uint16_t* localeName, int3
+{
+    assert(cwTargetLength >= 0);
+    Range result = {-2, 0};
+
+    NSLocale *currentLocale;
+    if (localeName == NULL || lNameLength == 0)
+    {
+        currentLocale = [NSLocale systemLocale];
+    }
+    else
+    {
+        NSString *locName = [NSString stringWithCharacters: localeName length: lNameLength];
+        currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:locName];
+    }
+
+    NSString *searchString = [NSString stringWithCharacters: lpTarget length: cwTargetLength];
+    NSString *searchStrComposed = ComposeString(searchString);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrComposed = ComposeString(sourceString);
+
+    if (searchStrComposed.length == 0)
+    {
+       result.location = fromBeginning ? 0 : sourceString.length;
+       return result;
+    }
+    NSRange rangeOfReceiverToSearch = NSMakeRange(0, sourceStrComposed.length);
+    NSStringCompareOptions options = ConvertFromCompareOptionsToNSStringCompareOptions(comparisonOptions);
+    
+    // in case mapping is not found
+    if (options == 0)
+        return result;
+    // last index
+    if (!fromBeginning)
+        options |= NSBackwardsSearch;
+    
+    NSRange nsRange = [sourceStrComposed rangeOfString:searchStrComposed
+                                         options:options
+                                         range:rangeOfReceiverToSearch
+                                         locale:currentLocale];
+    
+    if (nsRange.location != NSNotFound)
+    {   
+        result.location = nsRange.location;
+        result.length = nsRange.length;
+    }
+    else
+    {
+        result.location = -1;
+    }
+    
+    return result;
+}
+
+/*
+ Return value is a "Win32 BOOL" (1 = true, 0 = false)
+ */
+int32_t GlobalizationNative_StartsWithNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* lpPrefix, int32_t cwPrefixLength, 
+                                             const uint16_t* lpSource, int32_t cwSourceLength, int32_t comparisonOptions)
+                        
+{
+    NSLocale *currentLocale;
+    if(localeName == NULL || lNameLength == 0)
+    {
+        currentLocale = [NSLocale systemLocale];
+    }
+    else
+    {
+        NSString *locName = [NSString stringWithCharacters: localeName length: lNameLength];
+        currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:locName];
+    }
+
+    NSString *prefixString = [NSString stringWithCharacters: lpPrefix length: cwPrefixLength];
+    NSString *prefixStrComposed = ComposeString(prefixString);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrComposed = ComposeString(sourceString);
+
+    NSRange sourceRange = NSMakeRange(0, prefixStrComposed.length > sourceStrComposed.length ? sourceStrComposed.length : prefixStrComposed.length);
```

Comment:
I'm not sure if I agree. Why would the sourceRange depend in some cases on prefix/suffix? It should depend only on the source, especially when we use the sourceRange.Length as matchLenght value. `MatchLenght` answers the question: how many chars from source string are matched by prefix/suffix.
Example: 
`prefix = "\u00C0", source = "a\u0300"`
so prefix.Length = 1, source.Length = 2.
Expected match length: 2.
Actual range that we produce: 0...1.

Reply:
I'm using here `compare:options:range:locale:` which means I am taking range from `source` start till `prefix.length` and comparing with `prefix`. In case when `prefix` has longer length than `source` string `compare:options:range:locale:` throws exception. That is why check is there.

Reply:
no it doesn't , as `compare:options:range:locale:` is called for `source` string ,exception is only thrown if range is outside of it receivers sizes.
Your mentioned example will work fine as it will compare `a` with `"\u00C0"`.

Reply:
As initially strings are being `normalized to form C `( see https://developer.apple.com/documentation/foundation/nsstring/1412645-precomposedstringwithcanonicalma?language=objc) we will not have 2 same letters with different char lengths.